ams/ams-server/src/main/java/com/netease/arctic/ams/server/service/impl/TableExpireService.java

```diff
@@ -103,69 +101,94 @@ public void run() {
-            Set<String> baseExclude = UnKeyedTableUtil.getAllContentFilePath(changeTable);
-            baseExclude.addAll(finalHiveLocation);
-            expireSnapshots(baseTable, startTime - baseSnapshotsKeepTime, baseExclude);
-            long baseCleanedTime = System.currentTimeMillis();
-            LOG.info("[{}] {} base expire cost {} ms", traceId, arcticTable.id(), baseCleanedTime - startTime);
-
-            // delete ttl files
-            List<DataFileInfo> changeDataFiles = ServiceContainer.getFileInfoCacheService()
-                .getChangeTableTTLDataFiles(keyedArcticTable.id().buildTableIdentifier(),
-                    System.currentTimeMillis() - changeDataTTL);
-            deleteChangeFile(keyedArcticTable, changeDataFiles);
-
-            // get valid files in the base store which shouldn't physically delete when expire the snapshot
-            // in the change store
-            Set<String> changeExclude = UnKeyedTableUtil.getAllContentFilePath(baseTable);
-            changeExclude.addAll(finalHiveLocation);
-            expireSnapshots(changeTable, startTime - changeSnapshotsKeepTime, changeExclude);
-            return null;
-          });
-          LOG.info("[{}] {} expire cost total {} ms", traceId, arcticTable.id(),
-              System.currentTimeMillis() - startTime);
-        } else {
-          UnkeyedTable unKeyedArcticTable = arcticTable.asUnkeyedTable();
-          expireSnapshots(unKeyedArcticTable, startTime - baseSnapshotsKeepTime, hiveLocation);
-          long baseCleanedTime = System.currentTimeMillis();
-          LOG.info("[{}] {} unKeyedTable expire cost {} ms", traceId, arcticTable.id(), baseCleanedTime - startTime);
+    Set<String> hiveLocation = new HashSet<>();
+    if (TableTypeUtil.isHive(arcticTable)) {
+      hiveLocation = HiveLocationUtils.getHiveLocation(arcticTable);
+    }
+
+    if (arcticTable.isKeyedTable()) {
+      KeyedTable keyedArcticTable = arcticTable.asKeyedTable();
+      Set<String> finalHiveLocation = hiveLocation;
+      keyedArcticTable.io().doAs(() -> {
+        UnkeyedTable baseTable = keyedArcticTable.baseTable();
+        if (baseTable == null) {
+          LOG.warn("[{}] Base table is null: {} ", traceId, tableIdentifier);
+          return null;
         }
-      } catch (Throwable t) {
-        LOG.error("[" + traceId + "] unexpected expire error of table " + tableIdentifier, t);
+        UnkeyedTable changeTable = keyedArcticTable.changeTable();
+        if (changeTable == null) {
+          LOG.warn("[{}] Change table is null: {}", traceId, tableIdentifier);
+          return null;
+        }
+
+        // get valid files in the change store which shouldn't physically delete when expire the snapshot
+        // in the base store
+        Set<String> baseExclude = UnKeyedTableUtil.getAllContentFilePath(changeTable);
+        baseExclude.addAll(finalHiveLocation);
+        long latestBaseFlinkCommitTime = fetchLatestFlinkCommittedSnapshotTime(baseTable);
```

Comment:
Can you add some comment help other coder understand your logic here?

Reply:
OK, I added some comment for `fetchLatestFlinkCommittedSnapshotTime`.